File: nacos/3.2.3/analyzed/api/src/main/java/com/alibaba/nacos/api/naming/pojo/healthcheck/impl/Http.java
/*
 * Copyright 1999-2018 Alibaba Group Holding Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.alibaba.nacos.api.naming.pojo.healthcheck.impl;

import com.alibaba.nacos.api.common.Constants;
import com.alibaba.nacos.api.naming.pojo.healthcheck.AbstractHealthChecker;
import com.alibaba.nacos.api.utils.StringUtils;
import com.fasterxml.jackson.annotation.JsonIgnore;

import java.util.Collections;
import java.util.HashMap;
import java.util.Map;
import java.util.Objects;

/**
 * HTTP 协议健康检查器实现，向实例发起 HTTP 请求并根据响应码判定健康状态。
 *
 * <p>可配置探测路径、自定义请求头及期望 HTTP 状态码，默认期望 {@code 200}。</p>
 *
 * @author yangyi
 */
public class Http extends AbstractHealthChecker {
    
    /** 类型常量 {@code HTTP}。 */
    public static final String TYPE = "HTTP";
    
    private static final long serialVersionUID = 551826315222362349L;
    
    /** HTTP 探测路径，默认为空字符串。 */
    private String path = "";
    
    /** 原始请求头字符串，格式为 {@code Key:Value}，多条以分隔符连接。 */
    private String headers = "";
    
    /** 期望的 HTTP 响应状态码，默认 200。 */
    private int expectedResponseCode = 200;
    
    /** 构造 HTTP 类型健康检查器。 */
    public Http() {
        super(Http.TYPE);
    }
    
    /** 获取期望响应状态码。 */
    public int getExpectedResponseCode() {
        return this.expectedResponseCode;
    }
    
    /** 设置期望响应状态码。 */
    public void setExpectedResponseCode(final int expectedResponseCode) {
        this.expectedResponseCode = expectedResponseCode;
    }
    
    /** 获取 HTTP 探测路径。 */
    public String getPath() {
        return this.path;
    }
    
    /** 设置 HTTP 探测路径。 */
    public void setPath(final String path) {
        this.path = path;
    }
    
    /** 获取原始请求头字符串。 */
    public String getHeaders() {
        return this.headers;
    }
    
    /** 设置原始请求头字符串。 */
    public void setHeaders(final String headers) {
        this.headers = headers;
    }
    
    /**
     * 将 {@link #headers} 解析为键值 Map，供探测客户端使用。
     *
     * @return 请求头 Map；无有效头时返回空 Map
     */
    @JsonIgnore
    public Map<String, String> getCustomHeaders() {
        if (StringUtils.isBlank(headers)) {
            return Collections.emptyMap();
        }
        final Map<String, String> headerMap = new HashMap<>(16);
        for (final String s : headers.split(Constants.NAMING_HTTP_HEADER_SPLITTER)) {
            final String[] splits = s.split(":");
            if (splits.length != 2) {
                continue;
            }
            headerMap.put(StringUtils.trim(splits[0]), StringUtils.trim(splits[1]));
        }
        return headerMap;
    }
    
    @Override
    public int hashCode() {
        return Objects.hash(path, headers, expectedResponseCode);
    }
    
    @Override
    public boolean equals(final Object obj) {
        if (!(obj instanceof Http)) {
            return false;
        }
        
        final Http other = (Http) obj;
        
        if (!StringUtils.equals(path, other.getPath())) {
            return false;
        }
        if (!StringUtils.equals(headers, other.getHeaders())) {
            return false;
        }
        return expectedResponseCode == other.getExpectedResponseCode();
    }
    
    @Override
    public Http clone() throws CloneNotSupportedException {
        final Http config = new Http();
        config.setPath(getPath());
        config.setHeaders(getHeaders());
        config.setExpectedResponseCode(getExpectedResponseCode());
        return config;
    }
}

File: paddleocr/3.7.0/analyzed/api_sdk/typescript/src/internal/http.ts
// Copyright (c) 2026 PaddlePaddle Authors. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import {
  APIError,
  AuthError,
  InvalidRequestError,
  NetworkError,
  RateLimitError,
  RequestTimeoutError,
  ResponseFormatError,
  ResultParseError,
  ServiceUnavailableError,
} from "../errors.js";
import { userAbortReason } from "./abort.js";

// PaddleOCR 官方 API HTTP 客户端模块：提交任务、查询状态与拉取结果

const DEFAULT_BASE_URL = "https://paddleocr.aistudio-app.com";
const API_PATH = "/api/v2/ocr/jobs";

// 提交 OCR 任务时的可选参数
interface SubmitOptions {
  pageRanges?: string;
  batchId?: string;
  signal?: AbortSignal;
}

// 官方 API 统一响应包装：code/msg/data
interface APIResponse<T> {
  code?: number;
  msg?: string;
  data: T;
}

// 任务提交成功后的 jobId 响应体
interface SubmitResponse {
  jobId: string;
}

// HTTP 客户端：封装 Bearer 鉴权、超时、错误映射与 JSON/文件提交
export class HttpClient {
  private baseUrl: string;
  private jobsUrl: string;
  private token: string;
  private requestTimeout: number;
  private fetchImpl: typeof fetch;
  private clientPlatform?: string;

  // 初始化 baseUrl、jobs 端点、超时与 fetch 实现
  constructor(
    token: string,
    baseUrl: string = DEFAULT_BASE_URL,
    requestTimeout: number = 300000,
    fetchImpl: typeof fetch = fetch,
    clientPlatform?: string,
  ) {
    this.token = token;
    this.baseUrl = baseUrl.replace(/\/+$/, "");
    this.jobsUrl = `${this.baseUrl}${API_PATH}`;
    this.requestTimeout = requestTimeout;
    this.fetchImpl = fetchImpl;
    this.clientPlatform = clientPlatform;
  }

  // 通过远程 fileUrl 提交 OCR 任务，返回 jobId
  async submitUrl(model: string, fileUrl: string, optionalPayload: object, options: SubmitOptions = {}): Promise<string> {
    const body: Record<string, unknown> = { fileUrl, model, optionalPayload };
    if (options.pageRanges !== undefined) {
      body.pageRanges = options.pageRanges;
    }
    if (options.batchId !== undefined) {
      body.batchId = options.batchId;
    }
    const data = await this.fetchJson<SubmitResponse>(this.jobsUrl, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
    }, options.signal);
    return requireJobId(data);
  }

  // 通过本地文件 multipart 上传提交 OCR 任务
  async submitFile(model: string, filePath: string, optionalPayload: object, options: SubmitOptions = {}): Promise<string> {
    const fs = await import("fs");
    const path = await import("path");

    if (!fs.existsSync(filePath)) {
      const { FileNotFoundError } = await import("../errors.js");
      throw new FileNotFoundError(filePath);
    }

    const form = new FormData();
    form.append("model", model);
    form.append("optionalPayload", JSON.stringify(optionalPayload));
    if (options.pageRanges !== undefined) {
      form.append("pageRanges", options.pageRanges);
    }
    if (options.batchId !== undefined) {
      form.append("batchId", options.batchId);
    }

    const fileBuffer = fs.readFileSync(filePath);
    const blob = new Blob([fileBuffer]);
    form.append("file", blob, path.basename(filePath));

    const data = await this.fetchJson<SubmitResponse>(this.jobsUrl, {
      method: "POST",
      body: form,
    }, options.signal);
    return requireJobId(data);
  }

  // 查询单个异步任务的当前状态
  async getJobStatus(jobId: string, signal?: AbortSignal, timeoutMs?: number): Promise<unknown> {
    return this.fetchJson<unknown>(
      `${this.jobsUrl}/${encodeURIComponent(jobId)}`,
      { method: "GET" },
      signal,
      true,
      timeoutMs,
    );
  }

  // 查询批量任务下各子 job 的状态列表
  async getBatchStatus(batchId: string, signal?: AbortSignal, timeoutMs?: number): Promise<unknown> {
    return this.fetchJson<unknown>(
      `${this.jobsUrl}/batch/${encodeURIComponent(batchId)}`,
      { method: "GET" },
      signal,
      true,
      timeoutMs,
    );
  }

  // 下载 JSONL 格式结果并按行解析为对象数组
  async fetchJsonl(url: string, signal?: AbortSignal, timeoutMs?: number): Promise<unknown[]> {
    const resp = await this.fetch(url, { method: "GET" }, signal, false, timeoutMs);
    const text = await resp.text();
    try {
      return text
        .trim()
        .split("\n")
        .filter((line) => line.trim())
        .map((line) => JSON.parse(line) as unknown);
    } catch (error) {
      throw new ResultParseError("Failed to parse JSONL result payload.", { cause: error });
    }
  }

  // 下载二进制资源（图片、导出文件等）
  async fetchResource(url: string, signal?: AbortSignal, timeoutMs?: number): Promise<ArrayBuffer> {
    const resp = await this.fetch(url, { method: "GET" }, signal, false, timeoutMs);
    return resp.arrayBuffer();
  }

  // 内部 JSON 请求：校验 code/data 并反序列化
  private async fetchJson<T>(
    url: string,
    init: RequestInit,
    signal?: AbortSignal,
    withAuth: boolean = true,
    timeoutMs?: number,
  ): Promise<T> {
    const resp = await this.fetch(url, init, signal, withAuth, timeoutMs);
    let payload: APIResponse<T>;
    try {
      payload = await resp.json() as APIResponse<T>;
    } catch (error) {
      throw new ResponseFormatError("Expected a JSON response body.", { cause: error });
    }
    if (payload.code !== undefined && payload.code !== 0) {
      throw new APIError(resp.status, payload.msg || "PaddleOCR official API request failed.");
    }
    if (!payload || typeof payload !== "object" || !("data" in payload)) {
      throw new ResponseFormatError("Response body is missing data.");
    }
    return payload.data;
  }

  // 底层 fetch：合并鉴权头、AbortSignal 与 HTTP 状态码到异常
  private async fetch(
    url: string,
    init: RequestInit,
    signal?: AbortSignal,
    withAuth: boolean = true,
    timeoutMs?: number,
  ): Promise<Response> {
    const headers: Record<string, string> = {
      ...(init.headers as Record<string, string> || {}),
    };
    if (withAuth) {
      headers.Authorization = `Bearer ${this.token}`;
      if (this.clientPlatform) {
        headers["Client-Platform"] = this.clientPlatform;
      }
    }

    let resp: Response;
    const timeoutController = new AbortController();
    const effectiveTimeout = Math.max(0, Math.min(this.requestTimeout, timeoutMs ?? this.requestTimeout));
    const timeoutID = setTimeout(() => timeoutController.abort(), effectiveTimeout);
    const abortController = new AbortController();
    const abort = () => abortController.abort();
    timeoutController.signal.addEventListener("abort", abort, { once: true });
    if (signal?.aborted) {
      abort();
    } else {
      signal?.addEventListener("abort", abort, { once: true });
    }
    try {
      resp = await this.fetchImpl(url, {
        ...init,
        headers,
        signal: abortController.signal,
      });
    } catch (e: unknown) {
      if (signal?.aborted) {
        throw userAbortReason(signal);
      }
      if (timeoutController.signal.aborted) {
        throw new RequestTimeoutError(effectiveTimeout, { cause: e });
      }
      const message = e instanceof Error ? e.message : String(e);
      throw new NetworkError(`Connection failed: ${message}`);
    } finally {
      clearTimeout(timeoutID);
      signal?.removeEventListener("abort", abort);
    }

    if (resp.ok) return resp;

    let text = await resp.text();
    try {
      const payload = JSON.parse(text) as { msg?: string; message?: string; errorMsg?: string };
      text = payload.msg || payload.message || payload.errorMsg || text;
    } catch {
      // Keep raw response text.
    }
    if (resp.status === 401 || resp.status === 403) {
      throw new AuthError(`Authentication failed: ${text}`);
    } else if (resp.status === 400) {
      throw new InvalidRequestError(`Bad request: ${text}`);
    } else if (resp.status === 429) {
      throw new RateLimitError(`Rate limit exceeded: ${text}`);
    } else if (resp.status === 503 || resp.status === 504) {
      throw new ServiceUnavailableError(resp.status, `Service unavailable: ${text}`);
    } else {
      throw new APIError(resp.status, text);
    }
  }
}

// 校验提交响应必须包含非空 jobId
function requireJobId(data: SubmitResponse): string {
  if (!data || typeof data.jobId !== "string" || data.jobId.length === 0) {
    throw new ResponseFormatError("Submit response is missing jobId.");
  }
  return data.jobId;
}
